File: Commons/Graph/src/main/java/com/greentree/common/graph/algorithm/component/ConnectivityComponentsFinder.java
package com.greentree.common.graph.algorithm.component;

import java.util.Collection;
import java.util.List;

public interface ConnectivityComponentsFinder<V> {
	
	default int getComponent(V vertex) {
		final var cs = getComponents();
		for(int i = 0; i < cs.size(); i++) {
			final var c = cs.get(i);
			if(c.contains(vertex))
				return i;
		}
		throw new IllegalArgumentException("not vertex");
	}
	
	List<? extends Collection<V>> getComponents();
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/component/ConnectivityComponentsFinder.kt
package com.greentree.commons.graph.algorithm.component

interface ConnectivityComponentsFinder<out V> {

	fun getComponent(vertex: @UnsafeVariance V): Int {
		val cs = components
		for(i in cs.indices) {
			val c = cs[i]
			if(c.contains(vertex)) return i
		}
		throw IllegalArgumentException("not vertex")
	}

	val components: List<Collection<V>>
}
